Build images without Dapper

skitt opened the pull request:
We can achieve the same effect as running with Dapper, without Dapper,
by building images in multiple stages. The first stage runs in a
container built from our Shipyard base image, and subsequent stages
build our deliverable images.

This brings the following benefits:

* our container image builds are driven by our container descriptors,
  not by a combination of our Makefiles and container descriptors;
  this is more familiar to developers used to containers
* providing container descriptor-driven builds allows more third-party
  tools to be used directly, e.g. repository-hosted container image
  builds (on Quay or GitHub notably) and container descriptor analysis
  tools
* the risk of differences between "proper" container builds and local
  builds (with LOCAL_BUILD=1) is eliminated
* our build no longer depends on being able to build containers inside
  containers (which isn't supported everywhere)
* by building the binaries we want to ship in our container images
  during the container image build, we can ensure that the binaries
  are built for the architecture expected by the container image

The default target platform is linux/amd64 for the time being.

Depends on https://github.com/submariner-io/shipyard/pull/713

Signed-off-by: Stephen Kitt <skitt@redhat.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our developer guide: https://submariner.io/development/
2. Ensure you have added the appropriate tests for your PR: https://submariner.io/development/code-review/#test-new-functionality
3. Read the code review guide to ease the review process: https://submariner.io/development/code-review/
4. If the PR is unfinished, mark it as a draft: https://submariner.io/development/code-review/#mark-work-in-progress-prs-as-drafts
5. If you are using CI to debug, use your private fork: https://submariner.io/development/code-review/#use-private-forks-for-debugging-prs-by-running-ci
6. Add labels to the PR as appropriate.

This template is based on the K8s/K8s template:

https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

skitt committed:
Build images without Dapper

We can achieve the same effect as running with Dapper, without Dapper,
by building images in multiple stages. The first stage runs in a
container built from our Shipyard base image, and subsequent stages
build our deliverable images.

This brings the following benefits:

* our container image builds are driven by our container descriptors,
  not by a combination of our Makefiles and container descriptors;
  this is more familiar to developers used to containers
* providing container descriptor-driven builds allows more third-party
  tools to be used directly, e.g. repository-hosted container image
  builds (on Quay or GitHub notably) and container descriptor analysis
  tools
* the risk of differences between "proper" container builds and local
  builds (with LOCAL_BUILD=1) is eliminated
* our build no longer depends on being able to build containers inside
  containers (which isn't supported everywhere)
* by building the binaries we want to ship in our container images
  during the container image build, we can ensure that the binaries
  are built for the architecture expected by the container image

The default target platform is linux/amd64 for the time being.

Signed-off-by: Stephen Kitt <skitt@redhat.com>

dfarrell07 reviewed (dismissed):
LGTM

If anyone else is wondering about the repeating ARGs: https://docs.docker.com/engine/reference/builder/#understand-how-arg-and-from-interact

skitt commented:
To allow building in Buildah in rootless mode, this needs to run without Dapper at all, not just allow building without Dapper.

mkolesnik commented on package/Dockerfile.submariner-gateway:
```diff
@@ -1,4 +1,17 @@
+
+COPY . ${SOURCE}
+
+RUN make -C ${SOURCE} LOCAL_BUILD=1 bin/${TARGETPLATFORM}/submariner-gateway
```
Problem with multi-stage builds is inability to cache layers, i.e. this layer is not stored anywhere so it'll always have to be recreated, affecting any layers after it.

skitt replied:
We can use `--cache-to` with `mode=max` to cache all the layers.

skitt replied:
Note too that multi-stage builds are handled in a tree, not as a sequential list of layers. Layers in the final image are cached, as are layers from images explicitly loaded into the cache; so changes in one stage only affect layers in subsequent stage which depend on that stage (_i.e._ `COPY --from` lines and subsequent layers).

skitt replied:
I’m not sure caching _everything_ is desirable, since that would skip rebuilds in circumstances where we actually want a rebuild.

mkolesnik reviewed (commented):
This still leaves building the binaries outside the dapper as well, which leads to them being built twice on each change.

I also have a concern about caching.

skitt commented:
> This still leaves building the binaries outside the dapper as well, which leads to them being built twice on each change.

We can avoid that by dropping the `images: build` dependency; IIRC that causes issues elsewhere, which need some fixes in Shipyard, but it’s been a while so I’ve forgotten the details. I’ll look into it again.

skitt committed:
Add e2e dependency on vendor/modules.txt

Running e2e without building the binaries reveals that e2e needs
vendoring, so make that explicit by adding a dependency on
vendor/modules.txt.

Signed-off-by: Stephen Kitt <skitt@redhat.com>

skitt commented:
> > This still leaves building the binaries outside the dapper as well, which leads to them being built twice on each change.
> 
> We can avoid that by dropping the `images: build` dependency; IIRC that causes issues elsewhere, which need some fixes in Shipyard, but it’s been a while so I’ve forgotten the details. I’ll look into it again.

The problem is that `e2e` needs vendoring; I’ve added that as a separate commit.

dfarrell07 commented:
@sridhargaddam Can you take a look at this one?

mkolesnik commented:
I think the main question here is whats the benefit of doing this?

This seems to add a lot of complexity (and some degree of copy pasta) and perhaps some penalty on caching to the dockerfiles, but what are the benefits?

I feel the way we're building now is simple and also can be done without dapper anyhow with `LOCAL_BUILD` so I don't see that as a gain.

BTW we can remove `images: build` make dependency anyhow as it should be taken care of the Shipyard auto-dependencies (if that doesnt work we should fix it).

mkolesnik reviewed (commented):
I think the main question here is whats the benefit of doing this?

This seems to add a lot of complexity (and some degree of copy pasta) and perhaps some penalty on caching to the dockerfiles, but what are the benefits?

I feel the way we're building now is simple and also can be done without dapper anyhow with `LOCAL_BUILD` so I don't see that as a gain.

BTW we can remove `images: build` make dependency anyhow as it should be taken care of the Shipyard auto-dependencies (if that doesnt work we should fix it).

dfarrell07 commented:
> can be done without dapper anyhow with LOCAL_BUILD

Does that introduce more requirements for the configuration of a dev's machine, making the shared dev env less portable/sharable?

> but what are the benefits?

Not using Dapper seems like a clear benefit.

skitt commented:
See the updated PR description for a list of benefits (probably not exhaustive). See also the tech exchange where this was presented in context.

dfarrell07 commented:
@mkolesnik can you take a look at the benefits @skitt provided?

skitt committed:
Merge branch 'devel' into build-images-without-dapper
